### ISSUE TITLE
Install script fails with OPENCV4NODEJS_DISABLE_AUTOBUILD=1

### ISSUE THREAD

**Issue by Jayess** [justadudewhohacks/opencv4nodejs#183]
Here's the debug log from NPM attempt to install with AUTOBUILD off.

Ubuntu 16.04 LTS / OpenCV 3.2.0 build from source.

[npm_debug_log.txt](https://github.com/justadudewhohacks/opencv4nodejs/files/1761304/npm_debug_log.txt)

**Comment by justadudewhohacks** [comment]
Can not find anything helpful in the log. I am not using nvm, so not sure what's going on. Does nvm come bundled with node-gyp?

**Comment by Jayess** [comment]
I don't think it's bundled  with nvm.  It's installed with NPM, and NPM maintains a copy even if there's a -g install by the user.   Looking through the output, I see that the build fails for what appears to be a missing binding file.  I re-ran the failing part of the script using the ... node ... node-gyp.js ... rebuild command line.  Incidentally, there's no /home/jay/.node-gyp/8.9.4/<(target_arch)/node.lib' file existing or created.  Assumption is the arch folder should be x86_64.

Here's the output:

"/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp info it worked if it ends with ok
gyp info using node-gyp@3.6.2
gyp info using node@8.9.4 | linux | x64
gyp info spawn /usr/bin/python2
gyp info spawn args [ '/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/home/jay/Projects/saas.skustorm.com/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/home/jay/.node-gyp/8.9.4/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/home/jay/.node-gyp/8.9.4',
gyp info spawn args   '-Dnode_gyp_dir=/usr/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/home/jay/.node-gyp/8.9.4/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/home/jay/Projects/currentproject',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp: binding.gyp not found (cwd: /home/jay/Projects/currentproject) while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:336:16)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
gyp ERR! System Linux 4.13.0-36-generic
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/jay/Projects/currentproject
gyp ERR! node -v v8.9.4
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok

**Comment by justadudewhohacks** [comment]
Just curious, do you have python installed (preferably 2.7), which is a requirement for node-gyp. Also do install the package globally `npm i -g opencv4nodejs` or locally `npm i opencv4nodejs`?

**Comment by Jayess** [comment]
Yes.  Our dev systems have both 2.7 and 3.5 installed, with NPM config settings pointing to the 2.7 version, though the default already defaults to 2.7, and 3.5 must be explicitly configured or called.  Python behaves on our dev platforms exactly as it does with the 96 other packages we use in our project.

Your NPM install script seems runs to completion without the disable autobuild environment variable.  But it builds in non-default and nonstandard locations such that we couldn't rely on the avoidance of a future NPM rebuild or update wreaking havoc in a production system.  That's why we attempted the disable autobuild functionality in the first place.  It seems, though, that the nonstandard/nondefault install causes node-gyp, and possibly other reliant modules, to fail to locate the shared objects needed to fully install with autobuild disabled.

When CV2/CV3 builds are created in default locations (/usr/local/{share} and {bin}) test commands like "opencv_version" are executable from shell without any unusual pathing or environment variables being set.  In a standardized production environment we'd need to rely on that standard.  I don't think we can use this package under these conditions.  

Thanks for the work.  We do appreciate it.

**Comment by justadudewhohacks** [comment]
There is no need to install this package globally, I was just asking out of curiosity. Installing it locally is perfectly fine.

Having OpenCV set up globally in /usr/local and disabling the auto build via a flag is actually the preferable way to go, which is what I am doing on my local machine and what the CI systems are doing for example.

I have added the auto build feature in version 3.0 as I got quite some issues and questions from people, which had problems building OpenCV on their own, as against my expectations, a lot of people use opencv4nodejs to get started with OpenCV, without having lots of prior knowledge about OpenCV in general, let alone building C++ libraries with cmake from source on their own.

If for some reason this does not work on your system I am happy to help out of course. However, if you are providing me an npm log, which says `gyp ERR! configure error`, `gyp: binding.gyp not found ... while trying to load binding.gyp`,  I can not really say what might have potentially gone wrong. From this error message I would assume that there is something fundamentally going wrong when executing node-gyp on your system. Does the install script even execute at all? 

How do you infer: `fail to locate the shared objects or other build assets needed to properly install with autobuild disabled`? In case the install script doesn't find the OpenCV headers or libraries it would throw an error informing you about this. If there was an issue with linking the OpenCV libraries, node-gyp would inform you about this while building the package. If shared object libraries are not linked properly you would get an error at runtime e.g. `require('opencv4nodejs')` would fail with and error during process.dlopen.

Long story short, I would like to help you setting up the package, but it is hard for me to tell what the issue is if I can't reproduce it.

**Comment by Jayess** [comment]
Hi.
The failure of your install script to locate the build assets is not an inference, it's an observation.  Please check, above, the following output:

`gyp: binding.gyp not found (cwd: /home/jay/Projects/currentproject) while trying to load binding.gyp`

The behavior of your script when branched using disabled autobuild on our systems (more than one) is such that the called node-gyp process is unable to locate the (gyp) build assets required to complete the script and build the assets.  That's not personal.  It's just an observation about the operation of the script when so configured.

Your current CMAKE configuration also explicitly changes the install location on linux (which defaults to `/usr/local` see [here](https://cmake.org/cmake/help/v3.0/variable/CMAKE_INSTALL_PREFIX.html)).  Your script uses the following configuration, pulling paths from NPM's module location:
```
cmake [ '/home/jay/Projects/currentproject/node_modules/opencv-build/opencv/opencv',
  '-DCMAKE_INSTALL_PREFIX=/home/jay/Projects/currentproject/node_modules/opencv-build/opencv/build',
  '-DCMAKE_BUILD_TYPE=Release',
  '-DOPENCV_EXTRA_MODULES_PATH=/home/jay/Projects/currentproject/node_modules/opencv-build/opencv/opencv_contrib/modules',
  '-DBUILD_EXAMPLES=OFF',
  '-DBUILD_TESTS=OFF',
  '-DBUILD_PERF_TESTS=OFF',
  '-DBUILD_opencv_apps=OFF' ]

```
This change to the default install location (which impacts the ubiquity of command resources) modifies the defaults in such a way that they're not standardized to play well with other modules on platform that are configured to use OpenCVs standard install location.  Incidentally, you'll note that the CMAKE config above also explicitly switches off builds of the opencv_XXXXX commands (apps).  For all these reasons, we'd hoped to be able to use the disable autobuild option.  But the behavior of that script, when so branched, is documented above, and fails to complete.

**Comment by justadudewhohacks** [comment]
Hi,

I think `gyp: binding.gyp not found (cwd: /home/jay/Projects/currentproject) while trying to load binding.gyp` is just a confusing error message. Looking at your log:

```
1138 verbose lifecycle opencv4nodejs@3.1.0~install: CWD: /home/jay/Projects/currentproject/node_modules/opencv4nodejs
1139 silly lifecycle opencv4nodejs@3.1.0~install: Args: [ '-c', 'node-gyp rebuild' ]
1140 silly lifecycle opencv4nodejs@3.1.0~install: Returned: code: 1  signal: null
```

It says cwd is currentproject/node_modules/opencv4nodejs, which is where the binding.gyp file is located. So the process where node-gyp rebuild command is spawned is in the right location. Thus 'node-gyp rebuild' should not give the error `gyp: binding.gyp not found`, which is why I do not know why you get that error.

Regarding the cmake setting: When auto building the package it will build OpenCV inside the node_modules folder, so explicitly changing the cmake install directory to that path is intended. Also disabling the tests, examples and apps in the auto build is intended, as they are useless for this package anyways and just require unecessary build time. But if you are setting the disable auto build flag, this script is not being executed anyways, so you shouldn't bother about it.

**Comment by Jayess** [comment]
The script isn't working for us at all.  We've just tried to point out that the build, when executed using autobuild is located in a non-standard and non-default location despite your mention, above that:

> Having OpenCV set up globally in /usr/local and disabling the auto build via a flag is actually the preferable way to go, which is what I am doing on my local machine and what the CI systems are doing for example.

And when we disable autobuild the node-gyp package is unable to locate the build assets required to complete the build of your tools.  It's our observation that this is - likely - due to the nonstandard install locations for the build assets that are present in your script as assumptions about where assets should be located to complete the module builds required for your tools.

In both cases - the autobuild and the disabled autobuild - neither configuration works for us.  And, given the very standard configuration of our development environments (e.g. the default install location for OpenCV) it's very likely that our systems won't be the only ones that experience this.

Feel free to close this issue if you believe we can no longer provide information that can make your package more available in standardized installs and production environments.

Thanks!

**Comment by justadudewhohacks** [comment]
```
In both cases - the autobuild and the disabled autobuild - neither configuration works for us. And, given the very standard configuration of our development environments (e.g. the default install location for OpenCV) it's very likely that our systems won't be the only ones that experience this.
```

Well you are the first one coming up with this issue, maybe it has to do something with nvm.

What I am trying to say is the following: The error that you get are thrown **before** any of my install scripts are called. When the node-gyp process is spawned, it is 1. looking for a gyp file and 2. calling the scripts, which are specified in the gyp file, which include the headers and link the libraries.

The node-gyp process is not even spawning, which might call any scripts. Maybe you are better of asking in the node-gyp repo for help. I am just trying to help you, but all I can say is, I don't know why the node-gyp rebuild command fails for you.

**Comment by Jayess** [comment]
That could only make sense if node-gyp was failing for other modules/scripts.  It doesn't.  We use node-gyp (gyp) builds in 96 other modules for this project alone, and countless other modules in other projects.  Nothing in this scenario points to a failure of node-gyp or gyp configuration.

**Comment by justadudewhohacks** [comment]
Well the install script of opencv4nodejs is simply: `"install": "node-gyp rebuild"`. Maybe it fixes for you if you the execute node-gyp rebuild with the --directory flag poitning to cwd.

**Comment by Jayess** [comment]
Bear in mind, also, that the script runs to completion with no additional config changes to node-gyp or gyp when the autobuild is not disabled.  We pointed that out because it speaks against the presumption that there's a problem with gyp or node-gyp (or pre-gyp, for that matter).  

This is a problem with the script when autobuild is disabled.  And with autobuild left enabled, the install location for the opencv assets is non-standard, so unusable in our case.  It's a lose-lose scenario.

**Comment by justadudewhohacks** [comment]
Okay but the point is if you leave the flag enabled it will first install the opencv-build package (containing the scripts you mentioned). Does this build fail for you or is it just not feasible for you?

If it fails, then node-gyp rebuild of opencv4nodejs is not even called, so you would probably end up with the same problem here, even if the build succeeded.

But you are saying, that the install script of the opencv-build package finishes successfully and afterwards the node-gyp process for opencv4nodejs is spawned as expected? This behaviour would leave me totally confused.

**Comment by Jayess** [comment]
Yes.  That's what we're saying.  With autobuild in place (not disabled by env variable) the script runs, the cmake runs, the opencv build succeeds, the node-gyp build succeeds, and the package is installed.  With the variable in place and autobuild disabled, the package install fails as logged and attached in our first report.

**Comment by justadudewhohacks** [comment]
Okay that is completely odd. I have no clue why setting the flag would cause node-gyp to not find the binding.gyp anymore. Really my only explanation at this point is, that the install script gets executes and throws an error for some reason and the error message is being swallowed and instead node-gyp shows the error that the gyp file was not found.

Is it possible that for some reasons error message of errors thrown in an install script, do not get displayed in your log when running npm install?

**Comment by Jayess** [comment]
The logging sequence numbers (the first four numbers in the debug logs) seem to be sequential and contiguous.  So, no.  I'd say there's not much chance an error was thrown and not captured in the log.

**Comment by justadudewhohacks** [comment]
I am completely clueless what might be wrong then sorry. The only thing you could try out now is to clone this repository and npm install it from there. If this fails you could try to manually run node-gyp rebuild in the root directory and maybe pass the --directory flag to the root directory.

**Comment by Jayess** [comment]
We'll put up a docker environment or two to test a couple of different install scenarios once we get a couple of the current milestones completed. Then we'll be able to remove/reinstall OpenCV a few times trying different install scenarios for your package.  As it stands we have to have a solid OpenCV version installed in default to continue development.  So, it's not productive to keep removing/reinstalling to try to fix whatever is causing this.

I wouldn't put too much effort into this if you've reviewed the install scripts and think they're solid.  This could be something related to the permissions issues experienced on Linux boxes when using NPM and node-gyp, where ["root" doesn't have sufficient access](https://github.com/nodejs/node-gyp/issues/454) to certain directories.

That problem is corrected by using [Node Version Manager](https://docs.npmjs.com/getting-started/fixing-npm-permissions), which ensures that correct paths is applied to directory path builds and access.

While we're not saying they're related, we can say that we've seen behavior that isn't clearly intuitive from NPM and node-gyp before.   At the end of the day we're just looking for a working node interface for OpenCV, with testable, solid js code.  We may have to just write it ourselves, but you already understand the appeal of borrowing wheels that have already been invented.

**Comment by justadudewhohacks** [comment]
Hmm so I just tried with nvm and node v8.9.4 on ubuntu 14 and it works fine, so I guess nvm is not the issue.

This seems to be a common issue actually looking at the node-gyp issues, for example: https://github.com/nodejs/node-gyp/issues/233. Some people suggest clearing `.node-gyp` folder, maybe this helps.

**Comment by justadudewhohacks** [comment]
Closing this now. If your still after fixing this issue, feel free to reopen.

**Comment by droidMakk** [comment]
> There is no need to install this package globally, I was just asking out of curiosity. Installing it locally is perfectly fine.
> 
> Having OpenCV set up globally in /usr/local and disabling the auto build via a flag is actually the preferable way to go, which is what I am doing on my local machine and what the CI systems are doing for example.
> 
> I have added the auto build feature in version 3.0 as I got quite some issues and questions from people, which had problems building OpenCV on their own, as against my expectations, a lot of people use opencv4nodejs to get started with OpenCV, without having lots of prior knowledge about OpenCV in general, let alone building C++ libraries with cmake from source on their own.
> 
> If for some reason this does not work on your system I am happy to help out of course. However, if you are providing me an npm log, which says `gyp ERR! configure error`, `gyp: binding.gyp not found ... while trying to load binding.gyp`, I can not really say what might have potentially gone wrong. From this error message I would assume that there is something fundamentally going wrong when executing node-gyp on your system. Does the install script even execute at all?
> 
> How do you infer: `fail to locate the shared objects or other build assets needed to properly install with autobuild disabled`? In case the install script doesn't find the OpenCV headers or libraries it would throw an error informing you about this. If there was an issue with linking the OpenCV libraries, node-gyp would inform you about this while building the package. If shared object libraries are not linked properly you would get an error at runtime e.g. `require('opencv4nodejs')` would fail with and error during process.dlopen.
> 
> Long story short, I would like to help you setting up the package, but it is hard for me to tell what the issue is if I can't reproduce it.

I'm recieving a different issue.

> Error: Cannot find module '/home/pi/Documents/camstream-code/v1/opencv4nodejs/build/Release/opencv4nodejs'

**Comment by Piemontez** [comment]
The builder insists on recompiling opencv, even though I say OPENCV4NODEJS_DISABLE_AUTOBUILD = 1 = 0, using export using set.

Note: I am not adding "-w opencv4nodejs" after electron-rebuild

Because of that, I can't use opencv4 already installed on the system.

**Comment by Piemontez** [comment]
> The builder insists on recompiling opencv, even though I say OPENCV4NODEJS_DISABLE_AUTOBUILD = 1 = 0, using export using set.
> 
> Note: I am not adding "-w opencv4nodejs" after electron-rebuild
> 
> Because of that, I can't use opencv4 already installed on the system.

.... and a try add following settings to:

```
 "opencv4nodejs": {
    "disableAutoBuild": 1,
    "opencvIncludeDir": "/usr/include/opencv4",
    "opencvLibDir": "/usr/lib",
    "opencvBinDir": "/usr/lib"
  }
```
